File: backend/src/models/Department.js
'use strict';

const { Model, DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  class Department extends Model {
    static associate(models) {
      // Department has many members through department_members
      Department.belongsToMany(models.Member, {
        through: models.DepartmentMember,
        foreignKey: 'department_id',
        otherKey: 'member_id',
        as: 'members'
      });

      // Department has many department_members records
      Department.hasMany(models.DepartmentMember, {
        foreignKey: 'department_id',
        as: 'memberships'
      });

      // Department belongs to a leader (Member)
      Department.belongsTo(models.Member, {
        foreignKey: 'leader_id',
        as: 'leader'
      });

      // Self-referencing for hierarchy (parent-child departments)
      Department.belongsTo(models.Department, {
        foreignKey: 'parent_department_id',
        as: 'parentDepartment'
      });

      Department.hasMany(models.Department, {
        foreignKey: 'parent_department_id',
        as: 'subDepartments'
      });

      // Department has many meetings
      Department.hasMany(models.DepartmentMeeting, {
        foreignKey: 'department_id',
        as: 'meetings'
      });

      // Department has many tasks
      Department.hasMany(models.DepartmentTask, {
        foreignKey: 'department_id',
        as: 'tasks'
      });
    }
  }

  Department.init({
    id: {
      type: DataTypes.BIGINT,
      primaryKey: true,
      autoIncrement: true,
      allowNull: false
    },
    name: {
      type: DataTypes.STRING(150),
      allowNull: false,
      validate: {
        notEmpty: true,
        len: [2, 150]
      }
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    type: {
      type: DataTypes.STRING(50),
      allowNull: false,
      defaultValue: 'ministry',
      validate: {
        isIn: [['ministry', 'committee', 'service', 'social', 'administrative']]
      }
    },
    parent_department_id: {
      type: DataTypes.BIGINT,
      allowNull: true,
      field: 'parent_department_id'
    },
    leader_id: {
      type: DataTypes.BIGINT,
      allowNull: true,
      field: 'leader_id'
    },
    contact_email: {
      type: DataTypes.STRING(255),
      allowNull: true,
      validate: {
        isEmail: true
      }
    },
    contact_phone: {
      type: DataTypes.STRING(20),
      allowNull: true
    },
    meeting_schedule: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    is_active: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: true,
      field: 'is_active'
    },
    is_public: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: true,
      field: 'is_public'
    },
    max_members: {
      type: DataTypes.INTEGER,
      allowNull: true,
      validate: {
        min: 1
      }
    },
    sort_order: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0
    }
  }, {
    sequelize,
    modelName: 'Department',
    tableName: 'departments',
    timestamps: true,
    underscored: true
  });

  return Department;
};

File: backendJava/src/main/java/church/abunearegawi/backend/model/Department.java
package church.abunearegawi.backend.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "departments")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Department {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 150)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false, length = 50)
    @Builder.Default
    private String type = "ministry";

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_department_id")
    private Department parentDepartment;

    @OneToMany(mappedBy = "parentDepartment")
    private List<Department> subDepartments;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "leader_id")
    private Member leader;

    @Column(name = "contact_email", length = 255)
    private String contactEmail;

    @Column(name = "contact_phone", length = 20)
    private String contactPhone;

    @Column(name = "meeting_schedule", columnDefinition = "TEXT")
    private String meetingSchedule;

    @Column(name = "is_active", nullable = false)
    @Builder.Default
    private boolean isActive = true;

    @Column(name = "is_public", nullable = false)
    @Builder.Default
    private boolean isPublic = true;

    @Column(name = "max_members")
    private Integer maxMembers;

    @Column(name = "sort_order", nullable = false)
    @Builder.Default
    private Integer sortOrder = 0;

    @OneToMany(mappedBy = "department")
    private List<DepartmentMember> memberships;

    @OneToMany(mappedBy = "department")
    private List<DepartmentMeeting> meetings;

    @OneToMany(mappedBy = "department")
    private List<DepartmentTask> tasks;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum DepartmentType {
        ministry, committee, service, social, administrative
    }
}
